### PR TITLE
[QC-472] Extract advanced topology

### PR DESCRIPTION
**Pull request by Barthelemy**
Goal: be able to plug the QC with a different config file. 
I would like to have that to test the aggregators in particular. 

Also add documentation.

### CODE REVIEW

**Comment by Barthelemy** [comment]
@TimoWilken I am not sure what is going on but the tests are red while there are no details. Would you mind having a look when you have time ? thank you

**Comment by TimoWilken** [comment]
Hi @Barthelemy, I believe the build fails because there are conflicts between this PR and the master branch -- "Cannot merge PR into test area" is probably not the best message to show in this case, I'll try to think of a better one.

GitHub shows conflicts in `Framework/CMakeLists.txt`. Could you resolve that, e.g. by `git merge`-ing the latest master into your PR branch and fixing the conflict during that merge?

**Comment by Barthelemy** [comment]
Ok, totally missed the message, sorry about that :) I will merge manually

**Comment by Barthelemy** [comment]
@TimoWilken Hello, sorry to call again, it seems to me that o2-dataflow is not progressing.

File: Framework/include/QualityControl/AdvancedWorkflow.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+///
+
+#ifndef QUALITYCONTROL_AdvancedWorkflow_H
+#define QUALITYCONTROL_AdvancedWorkflow_H
+
+#include <Framework/CompletionPolicyHelpers.h>
+#include <Framework/ConfigContext.h>
+#include <Framework/DataSpecUtils.h>
+#include <Framework/CompletionPolicyHelpers.h>
+#include <Framework/WorkflowSpec.h>
```
Comment (knopers8) [inline Framework/include/QualityControl/AdvancedWorkflow.h:24]: You probably don't need most of these in the header.

File: Framework/src/AdvancedWorkflow.cxx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+#include <Framework/CompletionPolicyHelpers.h>
+#include <Framework/DataSpecUtils.h>
+#include "QualityControl/QcInfoLogger.h"
+#include <random>
+
+using namespace o2;
+using namespace o2::header;
+using namespace o2::framework;
+using SubSpecificationType = o2::header::DataHeader::SubSpecificationType;
+using SubSpec = o2::header::DataHeader::SubSpecificationType;
```
Comment (knopers8) [inline Framework/src/AdvancedWorkflow.cxx:28]: These two are the same, please pick one.

**Comment by TimoWilken** [comment]
@Barthelemy I'll have a look!

File: doc/ModulesDevelopment.md [hunk 3]
```diff
@@ -410,6 +410,21 @@ __Live detector data__
 
 If the detector is ready and connected to the CRU(s), one can of course start the full data taking workflow, including the SubTimeFrameBuilder and the DPL processing and plug the QC onto it.
 
+## A more advanced example
+
+A more complete example is available. The config file is called [advanced.json](../Framework/advanced.json). The workflow is made of 3 sources, intermediate processing steps, 3 sinks and a Dispatcher connecting two QC tasks to a number of these steps and 2 checks:
```
Comment (knopers8) [inline doc/ModulesDevelopment.md:415]: Please consider adding a sentence saying that the topology doesn't mean to represent any particular physics processing, it is just an example with multiple data processors.

**Review by knopers8 — COMMENTED** [review]
All good, I have just a few minor suggestions.

**Comment by Barthelemy** [comment]
Thank you , I have addressed all comments.

**Review by knopers8 — APPROVED** [review]

**Comment by Barthelemy** [comment]
The error on macOS is weird: `[ERROR] SEVERE: Device QC-TASK-RUNNER-skeletonTask (7203) had at least one message above severity ERROR: Unhandled std::exception reached the top of main: failed setting ZMQ_EVENTS, reason: Interrupted system call, device shutting down.`

I will trigger a rebuild

**Comment by Barthelemy** [comment]
@knopers8 any idea why `testWorkflow` is failing ? I cannot reproduce the problem on my mac and I don't see how my change can affect this test.

**Comment by knopers8** [comment]
No idea, but I've noticed that @ktf has been reworking integration with ZeroMQ in DPL, so it might be related.

**Comment by ktf** [comment]
Yes, it's definitely related to some of the reworking I was doing. Where do you get it?

**Comment by Barthelemy** [comment]
we get it in testWorkflow on Mac but only in this PR and I am not able to reproduce it on my Mac.
It happens when we quit: `        pctx.services().get<ControlService>().readyToQuit(QuitRequest::All);`

**Comment by Barthelemy** [comment]
seems to be fixed

**Comment by ktf** [comment]
@MichaelLettrich could you check what is going on with the macOS builder?

**Comment by pzhristov** [comment]
I another pull request I saw that the brew updated python to 3.9 and then the Python-modules-list was not OK, for example there was no matching version of tensorflow. A quick solution is to link python@3.8. If we want to use python@3.9, I have a patch that needs some refinement.

**Comment by MichaelLettrich** [comment]
@pzhristov @ktf indeed the macOS machines have been updated to python 3.9 via brew yesterday.  Do we need to downgrade them to 3.8?

**Comment by pzhristov** [comment]
@MichaelLettrich No need to downgrade, simply link python@3.8 
`brew link --force --overwrite python@3.8`

**Pull request merged**